### PR TITLE
WT-12172 Define the WT_RELEASE_WRITE macro which provides the write-release semantic

### PR DESCRIPTION
**Pull request by luke-pearson**

### CODE REVIEW

File: src/include/hardware.h [hunk 1]
```diff
@@ -19,6 +19,42 @@
...
+        } else if (sizeof((v)) == 2) {                              \
+            __asm__ volatile("stlrh %w1, %0" : "=Q"(v) : "r"(val)); \
+        } else if (sizeof((v)) == 4) {                              \
+            __asm__ volatile("stlr %w1, %0" : "=Q"(v) : "r"(val));  \
+        } else if (sizeof((v)) == 8) {                              \
+            __asm__ volatile("stlr %x1, %0" : "=Q"(v) : "r"(val));  \
+        }                                                           \
+    } while (0)
+#else
+#define WT_RELEASE_WRITE(v, val) __atomic_store_n((v), (val), __ATOMIC_RELEASE);
```
Comment (ajmorton) [inline src/include/hardware.h:52]:
```suggestion
#define WT_RELEASE_WRITE(v, val) __atomic_store_n(&(v), (val), __ATOMIC_RELEASE);
```

File: test/unittest/tests/test_acquire_release_macros.cpp [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    {                                                \
+        type a = 5;                                  \
+        type *ap = &a;                               \
+        type a_result;                               \
+        WT_ACQUIRE_READ(a_result, *ap);              \
+        REQUIRE(a == a_result);                      \
+                                                     \
+        type b;                                      \
+        type *bp = &b;                               \
+        type b_result = 5;                           \
```
Comment (ajmorton) [inline test/unittest/tests/test_acquire_release_macros.cpp:24]: Is this test passing? We never set the value of `b`

File: test/unittest/tests/test_acquire_release_macros.cpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        type a;                                      \
+        type *ap = &a;                               \
+        type a_set = value;                          \
+        WT_RELEASE_WRITE(*ap, a_set);                \
+        REQUIRE(a == a_set);                         \
+                                                     \
+        type b;                                      \
+        type *bp = &b;                               \
+        type b_set = value;                          \
+                                                     \
```
Comment (ajmorton) [inline test/unittest/tests/test_acquire_release_macros.cpp:41]: super nit:
```suggestion
```

File: test/unittest/tests/test_acquire_release_macros.cpp [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+ *
+ * This also helps verify the type checking in the acquire read macro, and that the barrier version
+ * doesn't produce different results.
+ */
+TEST_ACQUIRE_TYPE(uint64_t);
+TEST_ACQUIRE_TYPE(uint32_t);
+TEST_ACQUIRE_TYPE(uint16_t);
+TEST_ACQUIRE_TYPE(uint8_t);
+
+/* Test each branch of the release macro. */
```
Comment (ajmorton) [inline test/unittest/tests/test_acquire_release_macros.cpp:59]:
```suggestion
/* Test each branch of the release macro. Use values that can only fit inside the type being tested to make sure integer truncation doesn't occur. */
```

File: test/unittest/tests/test_acquire_release_macros.cpp [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+TEST_RELEASE_TYPE(uint16_t, UINT8_MAX + 1);
+TEST_RELEASE_TYPE(uint8_t, 1);
+
+/*
+ * We prevent users from releasing a value which isn't the same size as the type being released to.
+ * If unchecked it could result in integer overflow, both signed and unsigned. However this creates
+ * an issue when hash defined values are released, as they are assumed to be 8 bytes. We can work
+ * around this by casting in the hash define.
+ */
+TEST_CASE("Demonstrate hash define int size workaround", "[acqrel]")
```
Comment (ajmorton) [inline test/unittest/tests/test_acquire_release_macros.cpp:71]: Does this test fail if we don't cast TEST_VALUE? I'm also curious why we cast to uint when `a` is a signed int.

Comment (ajmorton) [reply test/unittest/tests/test_acquire_release_macros.cpp:71]: Additionally what does this mean for `#define`d values that don't have a typecast like
```
#define WT_PREPARE_INIT 0
#define WT_PREPARE_INPROGRESS 1
#define WT_PREPARE_LOCKED 2
#define WT_PREPARE_RESOLVED 3
```
?

File: src/include/hardware.h (hunk 1 above)
Comment (luke-pearson) [reply src/include/hardware.h]: ec9d34273

File: test/unittest/tests/test_acquire_release_macros.cpp (hunk 2 above)
Comment (luke-pearson) [reply test/unittest/tests/test_acquire_release_macros.cpp]: My guess is both were zero, this is C++ not C after all. ec9d34273

File: test/unittest/tests/test_acquire_release_macros.cpp (hunk 3 above)
Comment (luke-pearson) [reply test/unittest/tests/test_acquire_release_macros.cpp:41]: ec9d34273

File: test/unittest/tests/test_acquire_release_macros.cpp (hunk 4 above)
Comment (luke-pearson) [reply test/unittest/tests/test_acquire_release_macros.cpp]: ec9d34273

File: test/unittest/tests/test_acquire_release_macros.cpp (hunk 5 above)
Comment (luke-pearson) [reply test/unittest/tests/test_acquire_release_macros.cpp:74]: When we convert `WT_RELEASE_WRITE_WITH_BARRIER` to `WT_RELEASE_WRITE` we'll need to do the cast method. These #defines were what lead to me having to find that solution.
I added a comment to the test ec9d34273, explaining that it won't compile with the cast.

The cast may be ugly but it gives us correctness checking.

Comment (luke-pearson) [reply test/unittest/tests/test_acquire_release_macros.cpp:74]: We should keep this thread unresolved for others to see.

Comment (ajmorton) [reply test/unittest/tests/test_acquire_release_macros.cpp:74]: SGTM. I like the idea of added proper type information to `#define WT_PREPARE_INIT 0` in a future ticket.

File: test/unittest/tests/test_acquire_release_macros.cpp (hunk 3 above)
Comment (luke-pearson) [reply test/unittest/tests/test_acquire_release_macros.cpp]: bdde46842

File: src/support/hazard.c [hunk 6]
```diff
@@ -198,10 +198,10 @@ __wt_hazard_clear(WT_SESSION_IMPL *session, WT_REF *ref)
     for (hp = session->hazards.arr + session->hazards.inuse - 1; hp >= session->hazards.arr; --hp)
         if (hp->ref == ref) {
             /*
-             * We don't release write the hazard pointer clear in the general case. It's not
-             * required for correctness.
+             * Release write the hazard pointer. By doing so we guarantee that reads and writes of
+             * the pages memory cannot cross this point.
              */
-            hp->ref = NULL;
+            WT_RELEASE_WRITE(hp->ref, NULL);
```
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/support/hazard.c:204]: I would prefer this to be a separate ticket.

[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/support/hazard.c:204]: Sure, I'll take it out. The downside is WT_RELEASE_WRITE will need to be added to the s_defines list.

[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/support/hazard.c]: 3adf0a068

**Review by quchenhao — APPROVED** [review]
lgtm

**Review by ajmorton — APPROVED** [review]
LGTM 👍

**Pull request merged**